### ISSUE TITLE
Finalize Requirements

### ISSUE THREAD

**Issue by emanuelpalm** [arrowhead-f/x509-profile#1]
I was tasked by @jerkerdelsing before this summer to produce a proposal for a set of X.509 profiles for Arrowhead 5.0 and present them to you, as well as three others I do not have the GitHub usernames of. I will e-mail them to make sure that they see this discussion and the matter at hand.

However, before I sit down and write the proposal, I would like you to correct and/or ratify a list of requirements I have made regarding the profiles. The list of requirements is here: [`/requirements.md`](/arrowhead-f/x509-profiles/blob/main/requirements.md). If you find anything you want to add, remove or change, please indicate so by noting your change here.

The purpose of the profiles is to facilitate secure communication between Arrowhead systems within local clouds. Apart from enabling security and authentication, an important goal is also to facilitate authentication/access control. All of you have direct experience with different aspects of facilitating this. Mind you that the list of requirements is for Arrowhead 5.0. This is a chance to improve upon what exists, not an attempt to record and formalize it.

**Comment by emanuelpalm** [comment]
You will also, of course, be given a chance to comment or improve upon the final manuscript as well. This is just an attempt to align our thinking before I start, which could save me quite some time.

**Comment by emanuelpalm** [comment]
@mzsilak I am particularly interested in your thoughts regarding the manufacturer certificates. In your paper (https://doi.org/10.1109/ACCESS.2021.3102280), you only talk about one kind of manufacturer certificate, which is the one on the device that wants to be on-boarded. As far as I can tell, however, there needs to be two certificate types. The devices with these certificates need to have them issued by another certificate for them to become trustworthy.

I took the liberty of calling the issuing certificate for _Manufacturer Certificate_ and the issued certificate for _Transfer Certificate_ (the one you called Manufacturer Certificate in the paper). Do these certificates need to be RFC 5280-compliant (and thereby work with HTTPS), or is it enough for the transfer certificate to be trusted according to whatever X.509 profiles are relevant to the mode of communication? I would assume the latter. Am I wrong?

**Comment by tsvetlin** [comment]
Hello @emanuelpalm! Thank you for compiling a good requirements. We will also review it and post our comments. :)

**Comment by tsvetlin** [comment]
How are On-Boarding, Device, System, Operator distinguished between by Common Name?

**Comment by emanuelpalm** [comment]
@tsvetlin I have two proposals.

### 1. Use a Distinguished Name Qualifier (DNQ)

The Common Name (CN) is part of the `subject` field of the X.509 certificate, which in turn is a data structure of type `DistinguishedName` that contains one or more `RelativeDistinguishedName`s (see https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.4). The relative distinguished names form a hierarchy, akin to a DNS domain name, with the most significant relative name at the left and the most specific to the right (reverse order of DNS names). Each relative name contains an array of attribute type/value pairs, treated as if being at the same level of the hierarchy and, as a consequence, identifying the same entity. CN is one possible attribute type. Another is DNQ, which is meant to "qualify" other attribute values (quite vague, right?).

My fist proposal is to use the DNQ attribute type to indicate the entity type and the CN to only specify the entity name/identifier.

The following would be examples of possible `subject` field values (comma `,` separates relative distinguished names):
- `DNQ=system,CN=orchestrator-1`
- `O=My Company,DNQ=system,CN=serviceregistry-42`
- `O=My Company,DNQ=operator,CN=John Jones`

The `DNQ` field (called only `DN` in the softwares I have looked at) is positioned at the hierarchical level right above the name, which makes it akin to a conceptual folder containing entities of that type (such as systems or operators).

### 2. Separate Entity Type and Name With a Colon in the CN

This requires less explanation, but is, in my opinion, less beautiful as it requires the CN field to be parsed by the one who reads it.
A lot of software libraries are only loosely conformant to the X.509 specifications, however, (which is not strange as it is absolutely enormous), and a lot of times they only allow the CN field to be read directly. Relying more on the CN could help compatibility.

Examples corresponding to the above example `subject` fields would be:
- `CN=system:orchestrator-1`
- `O=My Company,CN=system:serviceregistry-42`
- `O=My Company,CN=operator:John Jones`

**Comment by emanuelpalm** [comment]
@jerkerdelsing @mzsilak Should we make recommendations or have requirements on the validity periods of certificates?

Also, have any of you, or anyone else, thought about how to handle the revocation of CA certificates? The Certificate Authority handles revocation of end entity certificates (system, operator, device and on-boarding certificates), but not any of the other ones. Should we mandate the use of OCSP (RFC 6960), like Firefox? Certificate revocation is and has been a bit of a hot topic, as there is no real consensus on the web about how to handle it. Chrome uses its own mechanism, for example. Neither Firefox nor Chrome uses the certificate revocation procedure outlined in RFC 5280, which defines the web X.509 certificate profile.

**Comment by jerkerdelsing** [comment]
@emanuelpalm @mzsilak I would say recommendations.

**Comment by jenseliasson** [comment]
Hi.
I completely agree with Jerker. Applications can be so diverse in
(security) requirements so recommendations is the only way.

It would also be interesting to produce a number of white papers with best
practices and recommendations for Arrowhead local clouds in different
scenarios.

Jens

**Comment by emanuelpalm** [comment]
@jerkerdelsing @jenseliasson I agree. And from you comment, @jenseliasson, I make out that we shouldn't put any specific recommendations in the Arrowhead X.509 Certificate Profiles document, but rather note that other documents may be published in the future by the Eclipse Arrowhead project.

**Comment by emanuelpalm** [comment]
@jerkerdelsing @jenseliasson @tsvetlin @mzsilak Should we disallow/recommend against human-selected CNs in certificates?

Humans tend to prefer understandable names, such as `pumping-station-04` or `serviceregistry-a`. Given that an adversary can get copies of some of these certificates, it may be able to deduce how many pumping stations exist in a given facility, or whether or not a local cloud has a service registry backup, etc. To protect against such information leaks, one could recommend that random identifiers be used instead. For example, a 12-character alphanumeric random identifier, such as `aiwm6qxa1vnx`.

Also DNS names and IP addresses in the subject alternative names extensions could leak this kind of information. They can be randomized in the same manner to avoid information leakage. IP addresses are a bit harder than DNS names, at least if IPv4 is used, as the possible address space is quite limited. The certificate type indicator I mentioned above also has the same problem, but it cannot be meaningfully randomized.

**Comment by jerkerdelsing** [comment]
@emanuelpalm I would say **recommend against** human selected CNs

**Comment by emanuelpalm** [comment]
Since most of you are yet to give more than a little input to this work, I've taken myself the freedom to make some decisions in the proposal that I've created. I will create a PULL request and refer to all of you in it later.

More specifically, I have assumed the following in the proposal:
- No manufacturer/transfer certificates are mentioned in this document. I figured they should be defined in a separate document, or be added later when we know more about them. @mzsilak 
- DN Qualifiers are used to indicate the type of certificate profile. @tsvetlin
- The Authorization system is assumed to be used within local clouds for certificate revocation. Outside them any other certificate revocation mechanism is allowed (i.e. we do not enforce any particular standard). As far as I can tell, this is the only thing that __must__ be discussed before we can ratify my proposal. We _should_ have some kind of mechanism for certificate revocation at all levels. @mzsilak

I'm closing this now. Further discussions should take place in the coming PR.
